False positive ``used-before-assignment`` on different line returns

This is related to my previous issue #4238 for the support of the walrus operator. Another edge case seems to be unhandled. That of different line returns.

### Steps to reproduce

Given a file `good-test.py`:
```python
"""Good test example."""
def func():
    """Test."""
    return f'The number {(count := 4)} is equal to {count}'

```
`pylint` doesn't detect any errors which is expected. However when return to a line like this:

Given a file `bad-test.py`:
```python
"""Bad test example."""
def func():
    """Test."""
    return f'The number {(count := 4)} ' \
           f'is equal to {count}'

```
We obtain the following:
### Current behavior

Result of `pylint bad-test.py`:
```
************* Module test
bad-test.py:4:19: E0601: Using variable 'count' before assignment (used-before-assignment)
```

### Expected behavior

Normally, this should not raise any errors.

### pylint --version output

Result of pylint --version output on my local macOS:

```
pylint 2.7.4
astroid 2.5.2
Python 3.8.8 (v3.8.8:024d8058b0, Feb 19 2021, 08:48:17) 
[Clang 6.0 (clang-600.0.57)]
```

--- a/ChangeLog
+++ b/ChangeLog
@@ -22,6 +22,10 @@ Release date: Undefined
   only contain ``__version__`` (also accessible with ``pylint.__version__``), other meta-information are still
   accessible with ``import importlib;metadata.metadata('pylint')``.
 
+* Fix false-positive ``used-before-assignment`` in function returns.
+
+  Closes #4301
+
 
 What's New in Pylint 2.7.5?
 ===========================

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1457,7 +1457,12 @@ class VariablesChecker(BaseChecker):
                             and defnode.lineno == node.lineno
                             and isinstance(
                                 defstmt,
-                                (astroid.Assign, astroid.AnnAssign, astroid.AugAssign),
+                                (
+                                    astroid.Assign,
+                                    astroid.AnnAssign,
+                                    astroid.AugAssign,
+                                    astroid.Return,
+                                ),
                             )
                             and isinstance(defstmt.value, astroid.JoinedStr)
                         )

--- a/tests/functional/a/assignment_expression.py
+++ b/tests/functional/a/assignment_expression.py
@@ -1,5 +1,6 @@
 """Test assignment expressions"""
-# pylint: disable=missing-docstring,unused-argument,unused-import,invalid-name,blacklisted-name,unused-variable
+# pylint: disable=missing-docstring,unused-argument,unused-import,invalid-name
+# pylint: disable=blacklisted-name,unused-variable,pointless-statement
 import re
 
 if (a := True):
@@ -14,6 +15,15 @@ x3 += b3 if (b3 := 4) else 6
 
 a = ["a   ", "b   ", "c   "]
 c = [text for el in a if (text := el.strip()) == "b"]
+
+
+# check wrong usage
+assert err_a, (err_a := 2)  # [used-before-assignment]
+print(err_b and (err_b := 2))  # [used-before-assignment]
+values = (
+    err_c := err_d,  # [used-before-assignment]
+    err_d := 2,
+)
 
 
 # https://github.com/PyCQA/pylint/issues/3347
@@ -53,7 +63,7 @@ print(function())
 
 
 # https://github.com/PyCQA/pylint/issues/3763
-foo if (foo := 3 - 2) > 0 else 0  # [pointless-statement]
+foo if (foo := 3 - 2) > 0 else 0
 
 
 # https://github.com/PyCQA/pylint/issues/4238
@@ -70,10 +80,7 @@ l3 += (
 )
 
 
-# check wrong usage
-assert err_a, (err_a := 2)  # [used-before-assignment]
-print(err_b and (err_b := 2))  # [used-before-assignment]
-values = (
-    err_c := err_d,  # [used-before-assignment]
-    err_d := 2,
-)
+# https://github.com/PyCQA/pylint/issues/4301
+def func2():
+    return f'The number {(count := 4)} ' \
+           f'is equal to {count}'

--- a/tests/functional/a/assignment_expression.txt
+++ b/tests/functional/a/assignment_expression.txt
@@ -1,4 +1,3 @@
-pointless-statement:56:0::Statement seems to have no effect
-used-before-assignment:74:7::Using variable 'err_a' before assignment
-used-before-assignment:75:6::Using variable 'err_b' before assignment
-used-before-assignment:77:13::Using variable 'err_d' before assignment
+used-before-assignment:21:7::Using variable 'err_a' before assignment
+used-before-assignment:22:6::Using variable 'err_b' before assignment
+used-before-assignment:24:13::Using variable 'err_d' before assignment

